### PR TITLE
Well-Architected Lens

### PR DESCRIPTION
**Pull request by jeremyber-aws**
Adding Managed Service for Apache Flink Well-Architected Lens

### CODE REVIEW

**Review by nicusX — COMMENTED** [review]
I added some minor comment.
I think there a couple of subjects that may be missing

1. serialization: make sure you are not falling back to Kryo, accidentally
2. What is the business impact if you lose the state of your application? Do you have a process to rebuild the state? This is not to cover accidents. You may be forced to do so to introduce incompatible changes

File: review/msf-well-architected.json [hunk 1]
```diff
@@ -0,0 +1,912 @@
...
+      "id": "dev_and_testing",
+      "name": "Development and Testing",
+      "questions": [
+        {
+          "id": "dev_and_testing_q1",
+          "title": "Have you developed your Apache Flink application locally and seen a successful execution of sample data?",
+          "description": "Developing locally is a critical piece of understanding how your application works, and if it will work well at scale.",
+          "helpfulResource": {
+            "displayText": "For more information in how to develop your Apache Flink applications locally, click the link provided",
+            "url": "https://docs.aws.amazon.com/managed-flink/latest/java/examples-getting-started-locally-flink.html"
```
[the remark this answers is not in the document]
Comment (jeremyber-aws) [reply review/msf-well-architected.json:16]: Added

File: review/msf-well-architected.json [hunk 2]
```diff
@@ -0,0 +1,912 @@
...
+          },
+          "choices": [
+            {
+              "id": "choice1",
+              "title": "Yes",
+              "improvementPlan": {
+                "displayText": "It is imperative that before pushing code to production, the connections to external data sources are functional. This can prevent a loss of data and other issues within an Apache Flink application."
+              }
+            },
+            {
```
[the remark this answers is not in the document]
Comment (jeremyber-aws) [reply review/msf-well-architected.json:65]: added

File: review/msf-well-architected.json [hunk 3]
```diff
@@ -0,0 +1,912 @@
...
+              "condition": "default",
+              "risk": "MEDIUM_RISK"
+            }
+          ]
+        },
+        {
+          "id": "dev_and_testing_q5",
+          "title": "Have you performance tested your application with realistic data mirroring production traffic, key cardinality and distribution?",
```
[the remark this answers is not in the document]
Comment (jeremyber-aws) [reply review/msf-well-architected.json:168]: rephrased

File: review/msf-well-architected.json [hunk 4]
```diff
@@ -0,0 +1,912 @@
...
+          "description": "In order to understand your application's performance, send realistic data through during performance tests to realistically simulate production traffic.",
+          "helpfulResource": {
+            "displayText": "You can use Managed Service for Apache Flink Studio, another Apache Flink application using the RichParallelSourceFunction, JMeter or another testing mechanism to simulate realistic traffic.",
```
[the remark this answers is not in the document]
Comment (jeremyber-aws) [reply review/msf-well-architected.json:171]: added

File: review/msf-well-architected.json [hunk 5]
```diff
@@ -0,0 +1,912 @@
...
+          "helpfulResource": {
+            "displayText": "Using lower level APIs like the ProcessFunction APIs for Apache Flink can lead to unexpected results and accumulation of state if not designed properly.",
+            "url": "https://nightlies.apache.org/flink/flink-docs-master/docs/dev/datastream/fault-tolerance/state/"
+          },
+          "choices": [
+            {
+              "id": "choice1",
+              "title": "Yes",
+              "improvementPlan": {
+                "displayText": "The Process Function APIs require great care and testing—if you need to manually manipulate state data, ensure you are effectively cleaning it out with StateTTLConfig, timers, and proper state management techniques."
```
[the remark this answers is not in the document]
Comment (jeremyber-aws) [reply review/msf-well-architected.json:253]: Changed to `.`

File: review/msf-well-architected.json [hunk 6]
```diff
@@ -0,0 +1,912 @@
...
+            },
+            {
+              "condition": "default",
+              "risk": "MEDIUM_RISK"
+            }
+          ]
+        },
+        {
+          "id": "dev_and_testing_q9",
+          "title": "Did you enable RocksDB metrics in development to monitor state access, size and latency?",
```
[the remark this answers is not in the document]
Comment (jeremyber-aws) [reply review/msf-well-architected.json:312]: Yeah I am not sure this is a reasonable expectation that customers will test this locally--am considering removing. WDYT?

File: review/msf-well-architected.json [hunk 7]
```diff
@@ -0,0 +1,912 @@
...
+              "title": "Yes",
+              "improvementPlan": {
+                "displayText": "N/A"
+              }
+            },
+            {
+              "id": "choice2",
+              "title": "No / I Don't Know",
+              "improvementPlan": {
+                "displayText": "Flink is optimized for high throughout and low latency, the logging subsystem is not. In case it is really required to generate log output for every processed message, use an additional DataStream inside the Flink application and a proper sink to send the data to Amazon S3 or CloudWatch. Do not use the Java logging system for this purpose. Moreover, Managed Service for Apache Flink’ Debug Monitoring Log Level setting generates a large amount of traffic, which can create backpressure. You should only use it while actively investigating issues with the application."
```
[the remark this answers is not in the document]
Comment (jeremyber-aws) [reply review/msf-well-architected.json:484]: For some reason it is backwards--You put the improvement plan on the option not selected. In this case, if the person says "Yes" they will see this text. I have tested and validated.

File: review/msf-well-architected.json [hunk 8]
```diff
@@ -0,0 +1,912 @@
...
+            },
+            {
+              "condition": "default",
+              "risk": "MEDIUM_RISK"
+            }
+          ]
+        },
+        {
+          "id": "monitoring_q4",
+          "title": "Have you implemented a mechanism to monitor end-to-end latency using custom metrics?",
```
[the remark this answers is not in the document]
Comment (jeremyber-aws) [reply review/msf-well-architected.json:505]: shall we nix this question, then?

**Comment by jeremyber-aws** [comment]
This PR also contains a CW Dashboard template in it now--we can split them if necessary. Requires review.

**Pull request closed without merging**